Wangscape/noise/module/ReseedableOps.cpp

```diff
@@ -0,0 +1,502 @@
+ReseedablePtr min(ReseedablePtr a, double b)
+{
+    ReseedablePtr c = makeConst(b);
+    return min(a, c);
+}
+
+ReseedablePtr min(double a, ReseedablePtr b)
+{
+    return min(b, a);
+}
+
+ReseedablePtr clamp(ReseedablePtr source, double lower, double upper)
+{
+    assert(lower <= upper);
+    // noise::module::Clamp cannot be used here because it asserts
+    // lower < upper
+    return max(lower, min(upper, source));
+}
+
+ReseedablePtr scaleBias(ReseedablePtr source, double scale, double bias)
+{
+    auto scale_bias_p = std::make_shared<Reseedable<ScaleBias>>();
+    scale_bias_p->module.SetSourceModule(0, source->getModule());
+    scale_bias_p->module.SetScale(scale);
+    scale_bias_p->module.SetBias(bias);
+
+    return makeModuleGroup({
+        {"source", source},
+        {DEFAULT_MODULE_GROUP_OUT, scale_bias_p}
+    });
+}
+
+ReseedablePtr terrace(ReseedablePtr source, int controlPointCount, bool inverted)
+{
+    auto terrace_p = std::make_shared<Reseedable<Terrace>>();
+    terrace_p->module.InvertTerraces(inverted);
+    terrace_p->module.MakeControlPoints(controlPointCount);
+    terrace_p->module.SetSourceModule(0, source->getModule());
+    return makeModuleGroup({
+        {"source", source},
+        {DEFAULT_MODULE_GROUP_OUT, terrace_p}
+    });
+}
+
+ReseedablePtr makeModuleGroup(std::initializer_list<std::pair<ModuleGroup::ModuleID, ReseedablePtr>> modules)
+{
+    auto module_group = std::make_shared<Reseedable<ModuleGroup>>();
+    for (auto name_module : modules)
+    {
+        module_group->module.insert(name_module.first, name_module.second);
+    }
+    return module_group;
+}
+
+ReseedablePtr selectQuadrant(ReseedablePtr source, bool x_positive, bool y_positive)
+{
+    if (x_positive && y_positive)
+        return source;
+    return translate(source, 
+                    x_positive ? 0. : -1.,
+                    y_positive ? 0. : -1.,
+                    0.);
+}
+
+ReseedablePtr makeCornerCombiner(bool x_positive, bool y_positive, double power)
+{
+    auto corner_combiner_base = std::make_shared<Reseedable<CornerCombinerBase>>();
+    corner_combiner_base->module.SetPower(power);
+    ReseedablePtr result = clamp(corner_combiner_base, -1, 1);
+    result = translate(result,
+                       x_positive ? 0. : -1.,
+                       y_positive ? 0. : -1.,
+                       0.0001); // Not a QuadrantSelector because of the z offset
+    return result;
+}
+
+ReseedablePtr makeEdgeFavouringMask(double p, double q, double min)
+{
+    auto norm_lp_q = std::make_shared<Reseedable<NormLPQ>>();
+    norm_lp_q->module.SetPQ(p, q);
+    ReseedablePtr result = translate(norm_lp_q, -1., -1., 0.);
+    result = scalePoint(result, 2., 2., 0.);
+    result = scaleBias(result, 1. - min, min);
+    result = clamp(result, min, 1.);
+    return result;
+}
+
+ReseedablePtr makeVariableScaleBias(ReseedablePtr source, ReseedablePtr min, ReseedablePtr max)
+{
+    return ((source + 1)*0.5*(max - min)) + min;
```

Comment:
These coefficients transform a `source` value in the range `[-1, 1]` to the range `[0, 1]`. It is then transformed to the range `[min, max]`.

I was considering making a module out of this so that users don't have to make 4 JSON modules to implement it. However, looking at the formula again, I think it's the same as `noise::module::Blend`, so that won't be necessary.